x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/mapper/preprocessor/MappingPreProcessor.java

```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.planner.mapper.preprocessor;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.xpack.esql.plan.logical.LogicalPlan;
+import org.elasticsearch.xpack.esql.plugin.TransportActionServices;
+
+/**
+ * Interface for a LogicalPlan processing rule occurring after the optimization, but before mapping to a physical plan.
+ * This step occurs on the coordinator. The rule may use services provided to the transport action and thus can resolve indices, rewrite
+ * queries, perform substitutions, etc.
+ * Note that the LogicalPlan following the rules' changes will not undergo another logical optimization round. The changes these rules
+ * should apply are only those that require access to services that need to be performed asynchronously.
+ */
+public interface MappingPreProcessor {
```

Comment:
Will there be multiple preprocessors?
Why not have a dedicated "stage"  such as preMapping() (similar to preAnalyze) and encapsulate the logic there?

Reply:
> Will there be multiple preprocessors?

That was my assumption, yes. 

> Why not have a dedicated "stage" such as preMapping() (similar to preAnalyze) and encapsulate the logic there?

There are some particularities to how queries are repeatedly rewritten (and stopping that is done by identity checking of a return object) -- these should remain particular to `FullTextFunction` IMO and this logic stay within a dedicated class. 

But the execution of the preprocessor is done at a dedicated stage.

Not sure if the current format meets your expectation, but happy to shape it further if it can be done better.